torch/_inductor/codegen/common.py

```diff
@@ -917,9 +975,72 @@ def inner(*args, **kwargs):
                 return inner
 
             @staticmethod
-            def indirect_indexing(index_var, size, check=True):
+            def indirect_indexing(var, size, check=True):
                 # Skip CSE since this doesn't return an expression
-                return self.indirect_indexing(index_var, size, check)  # type: ignore[attr-defined]
+
+                if var.bounds.lower < 0:
+                    new_bounds = ValueRanges.unknown()
+                    if var.bounds != ValueRanges.unknown() and isinstance(
+                        size, sympy.Number
+                    ):
+                        # Take the negative part of the bound and add size to it
+                        # Then take union of that and the positive part
+                        # This is a tighter bound than that of a generic ops.where, as we have info on the cond
+                        neg = var.bounds & ValueRanges(-sympy.oo, -1)
+                        new_bounds = ValueRanges(neg.lower + size, neg.upper + size)
+                        # We don't have a good way of representing the empty range
+                        if var.bounds.upper >= 0:
+                            pos = var.bounds & ValueRanges(0, sympy.oo)
+                            new_bounds = new_bounds | pos
+
+                    stm = ops.add(var, self.rename_indexing(size))
+                    # Mixed negative and non-negative
+                    if var.bounds.upper >= 0:
+                        lt = ops.lt(var, "0")
+                        stm = ops.where(lt, stm, var)
+                    new_var = self.cse.generate(self.compute, stm, bounds=new_bounds)
+
+                    new_var.update_on_args("index_wrap", (var,), {})
+                    var = new_var
+
+                if self.generate_assert(check):
+                    mask = ""
+                    if hasattr(var, "mask_vars"):
+                        mask_vars = set(var.mask_vars)
+                        if self._load_mask:
+                            mask_vars.add(self._load_mask)
+
+                        if mask_vars:
+                            mask = (
+                                f"{list(mask_vars)[0]}"
+                                if len(mask_vars) == 1
+                                else f"({' & '.join(str(v) for v in mask_vars)})"
+                            )
```

Comment:
All the mask logic is triton-only, so it should be in a per-backend aux function. For the same reason `_load_mask`  should just be in the triton kernel

Reply:
The cpp kernel seems to use `_load_mask` in some places as well. I'll keep it on `common.py` but move the code above to triton.